TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -101,107 +107,76 @@ - (TWTJSONSchemaTopLevelASTNode *)parseWithError:(NSError *__autoreleasing *)out
+        node = [[TWTJSONSchemaGenericASTNode alloc] init];
+        [self parseGenericSchema:schema intoNode:(TWTJSONSchemaGenericASTNode *)node withType:type];
     }
+
+    node.typeIsExplicit = typeIsExplicit;
+    [self parseCommonKeywordsFromSchema:schema intoNode:node];
+    return node;
 }
 
 
-- (TWTJSONSchemaArrayASTNode *)parseArraySchema:(NSDictionary *)arraySchema
+- (void)parseArraySchema:(NSDictionary *)arraySchema intoNode:(TWTJSONSchemaArrayASTNode *)node
 {
-    TWTJSONSchemaArrayASTNode *node = [[TWTJSONSchemaArrayASTNode alloc] init];
-    [self parseCommonKeywordsFromSchema:arraySchema intoNode:node];
-
     node.maximumItemCount = [self parseUnsignedIntegerForKey:TWTJSONSchemaKeywordMaxItems schema:arraySchema];
     node.minimumItemCount = [self parseUnsignedIntegerForKey:TWTJSONSchemaKeywordMinItems schema:arraySchema];
     node.requiresUniqueItems = [self parseBooleanForKey:TWTJSONSchemaKeywordUniqueItems schema:arraySchema valueIfNotPresent:NO];
 
-
     // "Items" must be JSON schema object or array of JSON schema objects
     if ([arraySchema[TWTJSONSchemaKeywordItems] isKindOfClass:[NSDictionary class]]) {
         node.itemSchemas = @[ [self parseSchemaForKey:TWTJSONSchemaKeywordItems schema:arraySchema] ];
+        node.itemsIsSingleSchema = YES;
         // additional items is meaningless if items is a single schema
     } else {
         node.itemSchemas = [self parseNonEmptyArrayOfSchemasForKey:TWTJSONSchemaKeywordItems schema:arraySchema];
+        node.itemsIsSingleSchema = NO;
         node.additionalItemsNode = [self parseAdditionalItemsOrPropertiesForKey:TWTJSONSchemaKeywordAdditionalItems schema:arraySchema];
     }
-
-    return node;
 }
 
 
-- (TWTJSONSchemaGenericASTNode *)parseGenericSchema:(NSDictionary *)genericSchema
+- (void)parseGenericSchema:(NSDictionary *)genericSchema intoNode:(TWTJSONSchemaGenericASTNode *)node withType:(NSString *)type
 {
-    TWTJSONSchemaGenericASTNode *node = [[TWTJSONSchemaGenericASTNode alloc] init];
-    [self parseCommonKeywordsFromSchema:genericSchema intoNode:node];
-
-    id type = genericSchema[TWTJSONSchemaKeywordType];
-    if ([type isKindOfClass:[NSArray class]]) {
-        node.validTypes = [NSSet setWithArray:type];
-    } else if (type) {
-        node.validTypes = [NSSet setWithObject:type];
-    }
-
-    return node;
+    node.validTypes = [NSSet setWithObject:type];
 }
 
 
-- (TWTJSONSchemaNumberASTNode *)parseNumberSchema:(NSDictionary *)numberSchema
+- (void)parseNumberSchema:(NSDictionary *)numberSchema intoNode:(TWTJSONSchemaNumberASTNode *)node
 {
-    TWTJSONSchemaNumberASTNode *node = [[TWTJSONSchemaNumberASTNode alloc] init];
-    [self parseCommonKeywordsFromSchema:numberSchema intoNode:node];
-
-    node.requireIntegralValue = [numberSchema[TWTJSONSchemaKeywordType] isEqualToString:TWTJSONSchemaTypeKeywordInteger];
+    node.requireIntegralValue = [numberSchema[TWTJSONSchemaKeywordType] isEqual:TWTJSONSchemaTypeKeywordInteger];
```

Comment:
In case `numberSchema[type]` is an array it would crash... but I've since updated it so this method will only ever get a string.